docker/archive/tar.go

```diff
@@ -0,0 +1,210 @@
+package archive
+
+// This implements all of the magical internal tar file handling.
+
+import (
+	"archive/tar"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+// fakeFile allows us to generate tar headers in-memory.
+type fakeFile struct {
+	path string
+	size int64
+}
+
+func (ff *fakeFile) Name() string {
+	return ff.path
+}
+func (ff *fakeFile) Size() int64 {
+	return ff.size
+}
+func (ff *fakeFile) Mode() os.FileMode {
+	return 0444
+}
+func (ff *fakeFile) ModTime() time.Time {
+	return time.Unix(0, 0)
+}
+func (ff *fakeFile) IsDir() bool {
+	return ff.Mode().IsDir()
+}
+func (ff *fakeFile) Sys() interface{} {
+	return nil
+}
+
+func sendTarFile(tw *tar.Writer, path string, size int64, stream io.Reader) error {
+	hdr, err := tar.FileInfoHeader(&fakeFile{
+		path: path,
+		size: size,
+	}, "")
+	if err != nil {
+		return err
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+	actualSize, err := io.Copy(tw, stream)
+	if err != nil {
+		return err
+	}
+	if size != actualSize {
+		return fmt.Errorf("send tar file: %s: mismatch between expected size (%d) and actual size (%d)", path, size, actualSize)
+	}
+	return nil
+}
+
+// tarCache allows for archive read operations to be sped up (and also
+// abstracted). It wraps an io.ReadSeeker with some helpful methods for dealing
+// with tar archives.
+type tarCache struct {
+	file    io.ReadSeeker
+	offsets map[string]int64
+}
+
+// prime regenerates all of the offset information in the tarCache.
+func (tc *tarCache) prime() error {
+	const tarBlockSize = 512
+
+	// Reset everything.
+	tc.offsets = map[string]int64{}
+	if _, err := tc.file.Seek(0, os.SEEK_SET); err != nil {
```

Comment:
Considering that `prime()` is called exactly once, immediately after creating the object, it would be simpler to call `prime()` first, and only then return a new object with the results of `prime()`; then the questions of “resetting” and “prior state” disappear.